File: 2015/day-16/day16.py
#!/usr/bin/python

from collections import defaultdict

MESSAGE = {"children": 3, "cats": 7, "samoyeds": 2, "pomeranians": 3, "akitas": 0,
            "vizslas": 0, "goldfish": 5, "trees": 3, "cars": 2, "perfumes": 1}

def main():
    sues = defaultdict(lambda: {})

    with open('input', 'r') as fp:
        for line in fp.read().replace(":", "").replace(",", "").split('\n'):
            i = line.split(" ")[1]
            line = line.split(" ")[2:]
            for j in range(int(len(line)/2)):
                sues[i][line[2*j]] = int(line[2*j+1])

    for i, gifts in sues.items():
        # part 1 checks
        possible = True
        for gift, amount in gifts.items():
            if amount != MESSAGE[gift]: 
                possible = False
                break
        if possible:
            print("part1: %s" % i)

        # part 2 checks
        possible = True
        for gift, amount in gifts.items():
            if gift in ["cats", "trees"]:
                if amount <= MESSAGE[gift]:
                    possible = False
                    break
            elif gift in ["pomeranians", "goldfish"]:
                if amount >= MESSAGE[gift]:
                    possible = False
                    break
            else:
                if amount != MESSAGE[gift]: 
                    possible = False
                    break
        if possible:
            print("part2: %s" % i)

if __name__ == '__main__':
    main()

File: CSharp/aoc/year_2015/day16/day16.cs
namespace Year2015.Day16;



public class Day16
{

    private static Dictionary<string, int> realSue = new Dictionary<string, int> { { "children", 3 }, { "cats", 7 }, { "samoyeds", 2 }, { "pomeranians", 3 }, { "akitas", 0 }, { "vizslas", 0 }, { "goldfish", 5 }, { "trees", 3 }, { "cars", 2 }, { "perfumes", 1 } };

    public static void Run(string Path)
    {
        string[] asf = File.ReadAllLines(Path).ToArray();
        int p1 = 0;
        int p2 = 0;

        foreach (string l in asf)
        {
            // Sue 88 trees 1 samoyeds 1 goldfish 0
            var aux = l.Replace(":", "").Replace(",", "").Split(" ");
            int sueNum = int.Parse(aux[1]);

            bool realP1 = true;
            bool realP2 = true;

            for (int i = 2; i < aux.Length; i += 2)
            {
                if (!realP1 && !realP2)
                    break;

                string compound = aux[i];
                int compoundVal = int.Parse(aux[(i + 1)]);

                if (compoundVal != realSue[compound])
                {
                    realP1 = false;
                    realP2 = false;
                }

                // p2 specific checks
                if (compound == "cats" || compound == "trees")
                    realP2 = compoundVal > realSue[compound] ? true : false;
                if (compound == "pomeranians" || compound == "goldfish")
                    realP2 = compoundVal < realSue[compound] ? true : false;

            }
            if (realP1 == true)
                p1 = sueNum;

            if (realP2 == true)
                p2 = sueNum;
        }
        Console.WriteLine($"Part 1: {p1}\nPart 2: {p2}");
    }

}